walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -56,55 +56,49 @@
+      if (areAttendeesInCurrentEvent && currentEvent.getWhen().start() >= startTime ) // Available time is only up to when event starts.
+      { 
+        currentTime = currentEvent.getWhen().start();
+
+        TimeRange potentialSlot = TimeRange.fromStartEnd(startTime, currentTime, false);
         if (potentialSlot.duration() >= requiredDuration)
         {
           candidateRanges.add(potentialSlot);
         }
-        break;
-      }
-
-      currentEvent = (Event)eventsIterator.next();
-
-      boolean areAttendeesInCurrentEvent = !Collections.disjoint(currentEvent.getAttendees(), request.getAttendees());
 
-      if (areAttendeesInCurrentEvent)
-      {
-        currentTime = currentEvent.getWhen().start(); 
-      } else 
-      {
-        continue; // We can take this timeslot as available and hence we look ahead to the next event.
-      }
-
-      //        start, current
-      //             v
-      // Timeline  : |--A-----|-------->
-      if (startTime == currentTime)
-      {
         startTime = currentEvent.getWhen().end();
-        currentTime = currentEvent.getWhen().end();
-        continue; //Look ahead to next event
-      }
-
-      //           start     current
-      //             v          v
-      // Timeline  : |----------|-----A-----|-------->
-      // Free time in between is potential range
-
-      TimeRange potentialSlot = TimeRange.fromStartEnd(startTime, currentTime, false);
-      if (potentialSlot.duration() >= requiredDuration)
+      }       
+      
+      if (eventsIterator.hasNext()) // Current event can be considered as available time. lookahead to next event.
+      {
+        currentEvent = (Event)eventsIterator.next();
+        if (previousEvent.getWhen().overlaps(currentEvent.getWhen()) && areAttendeesInCurrentEvent){ // Next event is intersecting and our start time needs to be set properly.
+          
+          // If the new event end is still during the previous event, we can set the new start time to the previous event's end, as that is a stricter bound.
+          // If the new event end is after the previous event, then we should consider the new event as per normal and start our timings from the start of the new event.
+          startTime = currentEvent.getWhen().end() < previousEvent.getWhen().end() ? previousEvent.getWhen().end() : currentEvent.getWhen().start();
+        }
+      } 
+      else // No events left, rest of time from startTime to end of day is free.
       {
-        candidateRanges.add(potentialSlot);
+        TimeRange potentialSlot = TimeRange.fromStartEnd(startTime, TimeRange.END_OF_DAY, true);
+        if (potentialSlot.duration() >= requiredDuration)
+        {
+          candidateRanges.add(potentialSlot);
+          startTime = TimeRange.END_OF_DAY;
+        }
+        break;
       }
-
-      startTime = currentEvent.getWhen().end();
-      currentTime = currentEvent.getWhen().end();
+      
     }
+
+    
```

Comment:
@123cristen @puthuparampil I refactored this so that my eventsIterator calls next at the end of the while loop rather than before. Lemme know if you need clarification or a rundown on how this works!